### ISSUE TITLE
TypeTag instances for SCrypt

### ISSUE THREAD

**Issue by daddykotex** [jmcardon/tsec#55]
I'm using Doobie and while trying to build an instance of `Meta` from another `Meta` using `xmap`, I've stumbed upon an issue. The `xmap` method needs a `TypeTag` type class instance. Usually those are provided by the compiler but in the case of `SCrypt`, the compiler was not able to do it.

I'm not familiar with `TypeTag` and I can't really help on this one. I worked around it by wrapping my `SCrypt` value inside a `case class TTSCrypt(value: SCrypt) extends AnyVal` and building a `Metat[TTSCrypt` instead.

It's not a major issue, it's just ugly.

Thanks for the library 👍 So far it works great!

**Comment by daddykotex** [comment]
I suspect the same thing happens for `BCrypt` and `HardenedSCrypt`

**Comment by jmcardon** [comment]
@daddykotex I know how to fix this issue quite easily, and it's something that requires library-wide changes (in a good way! nothing bad).

To explain a little, `SCrypt` is a little trick to use newtypes using `cats.evidence.Is`. In a sense, given that it's actually just a newtype, there is no type tag.

The actual definition for it in tsec is:

```scala
  protected val SCrypt$$ : TaggedString = new TaggedString {
    type I = String
    val is = Is.refl[String]
  }

  type SCrypt = SCrypt$$.I
```

This can give you compile-time newtypes that are typechecked, but can behave with much less overhead than case classes.

`is` has a `subst` method as such:
```scala
 /**
   * To create an instance of `A Is B` you must show that for every choice
   * of `F[_]` you can convert `F[A]` to `F[B]`. Loosely, this reads as
   * saying that `B` must have the same effect as `A` in all contexts
   * therefore allowing type substitution.
   */
  def substitute[F[_]](fa: F[A]): F[B]
```

Which means with the `is` instance for `SCrypt`, all you'd need to do is use `is.subst(meta)` to get a `Meta[SCrypt]`.

For now, you can use `metaString.asInstanceOf[Meta[SCrypt]]`, but I will make these fields public in the next milestone, thanks for pointing it out!

For the record, I've been enforcing this as of recently with my newtypes, as can be  seen here:

https://github.com/jmcardon/tsec/blob/tryswig/tsec-libsodium/src/main/scala/tsec/libsodium/hashing/package.scala#L27

Thank you! This solution is simple and simply using `SCrypt.is.subst(meta)` can fix it, but I didn't make the `is` public. I'll make sure to fix.

**Comment by jmcardon** [comment]
Also no, you don't need to wrap it in `AnyVal`. you can use `asInstanceOf[Meta[Scrypt]]` on a `Meta[String]` for now, and I'll make sure to fix this in a quick pr in a day or two (a tad busy today)

**Comment by daddykotex** [comment]
Oh wow. In the end, `SCrypt` is just a `String`.

And `Meta[String].asInstanceOf[Meta[SCrypt]]` worked just fine!

Should I close this?

**Comment by jmcardon** [comment]
@daddykotex not yet. I can pr a fix in a minute that's not ugly.

you shouldn't have to use `.asInstanceOf`. `Is.subst` is type checked and ultimately what you're after.

**Comment by daddykotex** [comment]
All right!

Also, there is no rush, take your time!

**Comment by jmcardon** [comment]
@daddykotex, not a problem, honestly, it was a quick fix that I intended to do anyway.

Thanks!

Fixed in #56, so I will be closing this issue. This fix will be out with the next milestone, which will release with Http4s-M6.

**Comment by daddykotex** [comment]
👍  thank you!
